runtime/sema/check_reference_expression.go

```diff
@@ -27,65 +27,12 @@ import (
 //
 func (checker *Checker) VisitReferenceExpression(referenceExpression *ast.ReferenceExpression) ast.Repr {
 
-	// Type-check the referenced expression
-
-	referencedExpression := referenceExpression.Expression
-
-	// Check that the referenced expression is an index expression and type-check it
-
-	var referencedType Type
-
-	// If the referenced expression is an index expression, it might be into storage
-
-	indexExpression, isIndexExpression := referencedExpression.(*ast.IndexExpression)
-	if isIndexExpression {
-		referencedType = checker.visitIndexExpression(indexExpression, false)
-
-		// The referenced expression will evaluate to an optional type if it is indexing into storage:
-		// the result of the storage access is an optional.
-		//
-		// Unwrap the optional one level, but not infinitely
-
-		if optionalReferencedType, ok := referencedType.(*OptionalType); ok {
-			referencedType = optionalReferencedType.Type
-		}
-
-		// Check if the index expression's target expression is a storage type
-
-	} else {
-		// If the referenced expression is not an index expression, check it normally
-
-		referencedType = referencedExpression.Accept(checker).(Type)
-	}
-
-	// Check that the references expression's type is not optional
-
-	getReferencedExpressionRange := func() ast.Range {
-		if isIndexExpression {
-			return ast.Range{
-				StartPos: indexExpression.TargetExpression.StartPosition(),
-				EndPos:   referencedExpression.EndPosition(),
-			}
-		} else {
-			return ast.NewRangeFromPositioned(referencedExpression)
-		}
-	}
-
-	if _, ok := referencedType.(*OptionalType); ok {
-
-		checker.report(
-			&OptionalTypeReferenceError{
-				ActualType: referencedType,
-				Range:      getReferencedExpressionRange(),
-			},
-		)
-	}
```

Comment:
I think we should keep this error, because it is a special case of referencing and the type mismatch error on its own might not be very helpful

Reply:
This is actually a bit tricky now. 
Since there is a type-mismatch error at the expression level, what we get back as the `referencedType` (i.e: the type of the referenced expr) is also the expected-type, not the actual type of the expression.

Even if we add that check/error back, we can only validate a scenario like:
```
let x: Int? = 1
let ref = &x as &Int?    // Note &Int?
```

But we can't validate this:
```
let x: Int? = 1
let ref = &x as &Int    // Note &Int
```

Reply:
We wouldn't still get the 'actual-type' of the referenced-expr.

I think we could solve this by making the `VisitExpression()` method to return two values (two types). 
- One is the 'actual-type' of the expression - This is not returned now.
- The type that others should see as the actual-type (like a 'visible-type') - This is what the method currently returns. 

For this particular case, we could use the 'actual-type'. Every other case would use the 'visible-type' for type checking - because it avoids the errors being delegated up.

Reply:
Added with abf2b1c

Reply:
Great idea, the commit looks good! 👍